### PR TITLE
Fix/Shard mode switch

### LINKED ISSUE #1956: Get rid of race-conditions with shard mode checking
Example: https://github.com/nspcc-dev/neofs-node/blob/master/pkg/local_object_storage/shard/gc.go#L181
Shard mode mutex should be taken on read for the whole operation.

### PR DESCRIPTION
**Pull request by carpawell**
Closes #1956.

### CODE REVIEW

File: pkg/local_object_storage/shard/gc.go [hunk 1]
```diff
@@ -214,9 +216,12 @@ func (s *Shard) removeGarbage() {
```
Comment (acid-ant) [inline pkg/local_object_storage/shard/gc.go:216]: Should we unlock here too?

**Review by acid-ant — APPROVED** [review]

File: pkg/local_object_storage/shard/gc.go (hunk 1 above)
Comment (carpawell) [reply pkg/local_object_storage/shard/gc.go:220]: for sure. rechecked all the places

File: CHANGELOG.md [hunk 2]
```diff
@@ -58,6 +58,7 @@ Changelog for NeoFS Node
 - Spawning useless `GETRANGE` with zero length for a big object (#2101)
 - Incomplete object put errors do contain the deepest error's message (#2092)
 - Prioritize internal addresses for clients (#2156)
+- Races on shard's mode switch (#1956)
```
Comment (fyrchik) [inline CHANGELOG.md:61]: It's race conditions, not races.

File: pkg/local_object_storage/shard/gc.go (hunk 1 above)
Comment (fyrchik) [reply pkg/local_object_storage/shard/gc.go:220]: Can we move all code to a separate function and use `defer` instead? It is ok to have 1-2 different `Unlock`, we have 4.

File: pkg/local_object_storage/shard/gc.go [hunk 3]
```diff
@@ -284,17 +298,25 @@ func (s *Shard) collectExpiredTombstones(ctx context.Context, e Event) {
 	for {
 		log.Debug("iterating tombstones")
 
-		if s.GetMode().NoMetabase() {
+		s.m.RLock()
```
Comment (fyrchik) [inline pkg/local_object_storage/shard/gc.go:301]: Same here.

File: CHANGELOG.md (hunk 2 above)
Comment (carpawell) [reply CHANGELOG.md]: fixed but what's the difference?

File: pkg/local_object_storage/shard/tree.go [hunk 4]
```diff
@@ -13,32 +13,44 @@ var ErrPiloramaDisabled = logicerr.New("pilorama is disabled")
...
 	}
 	return s.pilorama.TreeAddByPath(d, treeID, attr, path, meta)
 }
 
 // TreeApply implements the pilorama.Forest interface.
 func (s *Shard) TreeApply(d pilorama.CIDDescriptor, treeID string, m *pilorama.Move) error {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
 	if s.pilorama == nil {
```
Comment (fyrchik) [inline pkg/local_object_storage/shard/tree.go:49]: We can do this before locking a mutex.

File: pkg/local_object_storage/shard/gc.go (hunk 3 above)
Comment (carpawell) [reply pkg/local_object_storage/shard/gc.go:301]: `expiredTombstonesCallback` could call the same shard and deadlock once again

File: pkg/local_object_storage/shard/tree.go (hunk 4 above)
Comment (carpawell) [reply pkg/local_object_storage/shard/tree.go:46]: sure, just want to have the same logic as _the same as possible_ (lock in the same place (the first lines of the methods))

fixed

**Review by acid-ant — APPROVED** [review]

File: pkg/local_object_storage/shard/gc.go (hunk 1 above)
Comment (fyrchik) [reply pkg/local_object_storage/shard/gc.go:220]: You can split it into `delete` (no mutex) and `Delete` (take mutex + do `delete`).

Comment (carpawell) [reply pkg/local_object_storage/shard/gc.go]: @fyrchik, agree, fixed. unfortunately, i do not know how to solve that easily for our callbacks that are `engine` level config in fact. do you have any ideas?

**Review by fyrchik — APPROVED** [review]

**Pull request merged**